contracts/chain-adapters/Optimism_Adapter.sol

```diff
@@ -1,45 +1,63 @@
 // SPDX-License-Identifier: AGPL-3.0-only
 pragma solidity ^0.8.0;
 
+import "./Base_Adapter.sol";
+import "../interfaces/AdapterInterface.sol";
+import "../interfaces/WETH9.sol";
+
 import "@eth-optimism/contracts/libraries/bridge/CrossDomainEnabled.sol";
-import "@eth-optimism/contracts/L1/messaging/IL1ERC20Bridge.sol";
-import "./AdapterInterface.sol";
-import "@openzeppelin/contracts/access/Ownable.sol";
+import "@eth-optimism/contracts/L1/messaging/IL1StandardBridge.sol";
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 
 /**
  * @notice Sends cross chain messages Optimism L2 network.
  * @dev This contract's owner should be set to the BridgeAdmin deployed on the same L1 network so that only the
  * BridgeAdmin can call cross-chain administrative functions on the L2 SpokePool via this messenger.
  */
-contract Optimism_Messenger is Ownable, CrossDomainEnabled, AdapterInterface {
-    uint32 public gasLimit;
+contract Optimism_Adapter is Base_Adapter, CrossDomainEnabled {
+    uint32 public l2GasLimit = 5_000_000;
 
-    address l1Weth;
+    WETH9 l1Weth;
 
-    IL1ERC20Bridge l1ERC20Bridge;
+    IL1StandardBridge l1StandardBridge;
 
     constructor(
-        uint32 _gasLimit,
+        WETH9 _l1Weth,
+        address _hubPool,
         address _crossDomainMessenger,
-        address _IL1ERC20Bridge
-    ) CrossDomainEnabled(_crossDomainMessenger) {
-        gasLimit = _gasLimit;
-        l1ERC20Bridge = IL1ERC20Bridge(_IL1ERC20Bridge);
+        IL1StandardBridge _l1StandardBridge
+    ) CrossDomainEnabled(_crossDomainMessenger) Base_Adapter(_hubPool) {
+        l1Weth = _l1Weth;
+        l1StandardBridge = _l1StandardBridge;
     }
 
-    function relayMessage(address target, bytes memory message) external payable override onlyOwner {
-        sendCrossDomainMessage(target, uint32(gasLimit), message);
+    function setL2GasLimit(uint32 _l2GasLimit) public onlyOwner {
+        l2GasLimit = _l2GasLimit;
     }
 
-    // TODO: we should look into using delegate call as this current implementation assumes the caller
-    // transfers the tokens first to this contract.
+    function relayMessage(address target, bytes memory message) external payable override onlyHubPool {
```

Comment:
we might as well emit events in both `relayMessage` and `relayTokens`? Especially `relayTokens`

Reply:
can.